pino-opentelemetry-transport.js

```diff
@@ -0,0 +1,136 @@
+ * @property {string | number} destination
+ * @property {string} [messageKey="msg"]
+ *
+ * @param {Options} opts
+ */
+module.exports = async function (opts) {
+  const destination = new SonicBoom({ dest: opts.destination, sync: false })
+  const mapperOptions = {
+    messageKey: opts.messageKey || DEFAULT_MESSAGE_KEY
+  }
+
+  return build(async function (/** @type { AsyncIterable<Bindings> } */ source) {
+    for await (const obj of source) {
+      const updatedLine = JSON.stringify(toOpenTelemetry(obj, mapperOptions)) + '\n'
+      const writeResult = destination.write(updatedLine)
+      const toDrain = !writeResult
+      // This block will handle backpressure
+      if (toDrain) {
+        await once(destination, 'drain')
+      }
+    }
+  }, {
+    async close () {
+      destination.end()
+      await once(destination, 'close')
+    }
+  })
+}
+
+const FATAL_SEVERITY_NUMBER = 21
+/**
+ * If the source format has only a single severity that matches the meaning of the range
+ * then it is recommended to assign that severity the smallest value of the range.
+ * https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#mapping-of-severitynumber
+ */
+const SEVERITY_NUMBER_MAP = {
+  10: 1,
+  20: 5,
+  30: 9,
+  40: 13,
+  50: 17,
+  60: FATAL_SEVERITY_NUMBER
+}
+
+// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#displaying-severity
+const SEVERITY_NAME_MAP = {
+  1: 'TRACE',
+  2: 'TRACE2',
+  3: 'TRACE3',
+  4: 'TRACE4',
+  5: 'DEBUG',
+  6: 'DEBUG2',
+  7: 'DEBUG3',
+  8: 'DEBUG4',
+  9: 'INFO',
+  10: 'INFO2',
+  11: 'INFO3',
+  12: 'INFO4',
+  13: 'WARN',
+  14: 'WARN2',
+  15: 'WARN3',
+  16: 'WARN4',
+  17: 'ERROR',
+  18: 'ERROR2',
+  19: 'ERROR3',
+  20: 'ERROR4',
+  21: 'FATAL',
+  22: 'FATAL2',
+  23: 'FATAL3',
+  24: 'FATAL4'
+}
+
+/**
+ * Converts a pino log object to an OpenTelemetry log object.
+ *
+ * @typedef {Object} OpenTelemetryLogData
+ * @property {string=} SeverityText
+ * @property {string=} SeverityNumber
+ * @property {string} Timestamp
+ * @property {string} Body
+ * @property {{ 'host.hostname': string, 'process.pid': number }} Resource
+ * @property {Record<string, any>} Attributes
+ *
+ * @typedef {Object} MapperOptions
+ * @property {string} messageKey
+ *
+ * @param {Bindings} sourceObject
+ * @param {MapperOptions} mapperOptions
+ * @returns {OpenTelemetryLogData}
+ */
+function toOpenTelemetry (sourceObject, { messageKey }) {
+  const { time, level, hostname, pid, [messageKey]: msg, ...attributes } = sourceObject
+
+  const severityNumber = SEVERITY_NUMBER_MAP[sourceObject.level]
+  const severityText = SEVERITY_NAME_MAP[severityNumber]
+
+  return {
+    Body: msg,
+    Timestamp: time + ZEROS_FROM_MILLI_TO_NANO,
+    SeverityNumber: severityNumber,
+    SeverityText: severityText,
+    Resource: {
+      'host.hostname': hostname,
+      'process.pid': pid
+    },
+    Attributes: attributes
```

Comment:
Fastify and pino-http provides a `reqId` property that is unique for a given request. It might be nice to map that to a trace or a span.

We might want to have some logic to decide if an http request was completed. I'd be happy to add some property in the logs of Fastify/pino-http for this purpose or maybe we could just match the message text (and possibly make this configurable).

Reply:
Ok, I was thinking of adding a custom mapper object to this transport's options. The user could then decide how they want their props mapped. The mapper would override the default behaviour.